### PR TITLE
@types/node: Added info object result to Zlib functions.

### PR DESCRIPTION
**Pull request by WilcoBakker**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: See `info` option [here](https://nodejs.org/api/zlib.html#class-options). (It states that if true, that it returns an object with buffer and engine.)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

Albeit this only being a fix as TypeScript is only layer a top of JavaScript and TypeScript didn't reflect the JavaScript underneath it properly in this case, it is not out of the question though that some TypeScript code might break because of this so this is a breaking change.  My last contribution was 6 years ago and this contribution is significantly larger. Any other feedback would be appreciated as well.

### CODE REVIEW

**Comment by Semigradsky** [comment]
Please, also `v16/ts4.8` and `v18/ts4.8`

**Comment by WilcoBakker** [comment]
Yes, thank you, @Semigradsky, I will add those too.

**Comment by WilcoBakker** [comment]
Now those are added as well! ^^ @Semigradsky

**Comment by Semigradsky** [comment]
Looks like need to change `ZlibOptions` here to `ZlibOptionsBase`:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/inflation/index.d.ts#L9C10-L9C21

File: types/inflation/index.d.ts
Comment (Semigradsky) [inline types/inflation/index.d.ts:21]: `interface Options extends ZlibOptionsBase {` is not enough?

Comment (WilcoBakker) [reply types/inflation/index.d.ts:21]: Practically speaking? Perhaps. I believe the `info` option only does something when used with deflate, deflateRaw, gzip, gunzip, inflate, inflateRaw or unzip but it would deviate from the official Node.JS documentation (it doesn't clarify whether the `info` option is ignored for the classes or not) and the TypeScript definitions so that is the reason why I added both flavors.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/node/zlib.d.ts
1. types/inflation/index.d.ts
1. types/node/v18/zlib.d.ts
1. types/node/ts4.8/zlib.d.ts
1. types/node/v16/zlib.d.ts

<details>
<summary>as well as these 8 other files...</summary>
<p>types/node/test/zlib.ts, types/node/v18/ts4.8/zlib.d.ts, types/node/v18/test/zlib.ts, types/node/ts4.8/test/zlib.ts, types/node/v16/ts4.8/zlib.d.ts, types/node/v16/test/zlib.ts, types/node/v18/ts4.8/test/zlib.ts, types/node/v16/ts4.8/test/zlib.ts</p>
</details>


Consider running `npx dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 100c8e3b82af354f2a8ee45e2c3b25b878cd1123
</p>

**Comment by WilcoBakker** [comment]
@Semigradsky where should I bump up the versions of @types/node and @types/inflation by the way? Because that is the only thing I didn't do yet.

**Comment by Semigradsky** [comment]
They will be upped automatically. Now just waiting for approve from code owners.

**Comment by WilcoBakker** [comment]
Thank you @Semigradsky, Could you perhaps approve your review by the way? Assuming you agree with my argumentation regarding your comment.

Also, on a side note, would you know how I could generate the npm package of @types/node myself? Because while waiting for the code owners because then I can at least use the changes I made myself while wating for the code owners to approve my changes.

File: types/node/zlib.d.ts
Comment (WilcoBakker) [reply types/node/zlib.d.ts:121]: I would beg to differ because leaving `ZlibOptions` as is would mean that the overloads I made won't work. Because when the `info` property of the `ZlibOptions` stays the same, `ZlibOptionsWithInfo` would remain backwards-compatible which won't play well with the overloads. The lack of the (correct) implementation of the `info` property unfortunately means that a fix would be impactful. At the end of the day you unfortunately have to work with overloads as the functions of the `Zlib` library have 2 different return types based on the `info` property of the `ZlibOptions`.

Could you see my point of view?

Comment (remcohaszing) [reply types/node/zlib.d.ts:121]: TypeScript picks the first matching overload. If the overload using `ZlibOptionsWithInfo` goes first, it takes precedence and the callback / return values are inferred the way I think you intend to happen.

You can see this in the playground link if you hover the `result` arguments in the callbacks.

Comment (WilcoBakker) [reply types/node/zlib.d.ts:121]: I don't think that is a very solid design strategy. Not to mention that I'd doubt it would be much more less impactful because it would remain a breaking change because the return type does still change from `Buffer` to the info object when `info` equals `true`.

File: types/node/ts4.8/zlib.d.ts
Comment (jakebailey) [inline types/node/ts4.8/zlib.d.ts:121]: This change in declaration is a breaking change; is there a downside to leaving this behind as the base instead?

File: types/inflation/index.d.ts
Comment (WilcoBakker) [reply types/inflation/index.d.ts]: By second thought, I reverted changes to the `ZlibOptions`  interface, my proposed change of it being an union type of my two new introduced interfaces 'ZlibOptionsWithoutInfo` and `ZlibOptionsWithInfo` has hereby made undone, making changes to the `@types/inflation` package no longer necessary and thus have been reverted as well so this PR no longer affects it.

File: types/node/zlib.d.ts
Comment (WilcoBakker) [reply types/node/zlib.d.ts]: I have made a change, `ZlibOptions` interface has been reinstated in full glory and it no longer is a union type of my two newly introduced interfaces. Changes to the `@types/inflation` package has been reverted by this as well and the package is now unaffected.

File: types/node/ts4.8/zlib.d.ts
Comment (WilcoBakker) [reply types/node/ts4.8/zlib.d.ts]: I understand it is a breaking change, just now made some changes to so that the `ZlibOptions` interface itself is now unaffected, unfortunately this PR does remain a breaking change but that is just because the initial definition just didn't include the info object that the Node.JS documentation describes which it should have.

**Comment by WilcoBakker** [comment]
Just made a change that reinstates the `ZlibOptions` interface and leaving it unaffected in favor of it becoming a union type of my two newly introduced interfaces `ZlibOptionsWithoutInfo` and `ZlibOptionsWithInfo`. Unfortunately this doesn't prevent this PR being a breaking change but it does reduce the impact of this change, I just don't see any other way to prevent this not being a breaking change.

File: types/node/zlib.d.ts
Comment (jakebailey) [reply types/node/zlib.d.ts]: Putting the more specific overload first is the normal way to write overloads in TypeScript; what's not a solid design strategy? I would think there's precedent when it comes to say, `child_process`'s return values that could be used to model the API here.

Comment (WilcoBakker) [reply types/node/zlib.d.ts]: Okay, thanks for the clarification on that part. The main take away would be that it doesn't prevent this from being a breaking change though.

**Comment by WilcoBakker** [comment]
I removed `ZlibOptionsWithoutInfo` in favor of precedence in overloads as advised by @remcohaszing.

**Comment by jakebailey** [comment]
Theoretically this looks good to me, but I think I'd prefer if someone on the node types owner list would look.

I'll note that it's somewhat suspicious to have overloads that are entirely differing types at the return position, but without a final "return type big union" overload that is the catchall for when the input could match either of the other overloads. But, maybe TS just synthesizes that overload automatically?

The closes analogue to this change IIRC is how `child_process` handles its output type between Buffer and string, so that'd be the place to verify that it's consistent, as my feeling is that those overloads have been working well for a very long time.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
we would want review from one of the owners to see if this is correct

**Comment by WilcoBakker** [comment]
@sheetalkamat could you please resolve your request for changes? I assume you did it by mistake and now it says I need to do a revision which doesn't help this PR.

I would very kindly request one of the code owners ( @eps1lon and @peterblazejewicz ) to take a look at this PR and provide me with feedback, it would be very much appreciated!

**Pull request merged**
